drivers/serial/uart_nrf_sw_lpuart.c

```diff
@@ -591,17 +825,59 @@ static int lpuart_init(struct device *dev)
 	k_timer_init(&data->tx_timer, tx_timeout, NULL);
 	k_timer_user_data_set(&data->tx_timer, dev);
 
-	return uart_callback_set(data->uart, uart_callback, (void *)dev);
+	err = uart_callback_set(data->uart, uart_callback, (void *)dev);
+	if (err < 0) {
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_NRF_SW_LPUART_INT_DRIVEN)) {
+		err = uart_callback_set(dev, int_driven_evt_handler, dev);
+		if (err < 0) {
+			return -EINVAL;
+		}
+
+		err = api_rx_enable(dev, data->int_driven->rxbuf,
+				    sizeof(data->int_driven->rxbuf), 1);
+	}
+
+	data->txbyte = -1;
+
+	return err;
 }
 
 static int api_poll_in(struct device *dev, unsigned char *p_char)
 {
-	return -ENOTSUP;
+	if (IS_ENABLED(CONFIG_NRF_SW_LPUART_INT_DRIVEN)) {
+		return api_fifo_read(dev, p_char, 1) ? 0 : -1;
+	}
+
+	return -EBUSY;
 }
 
 static void api_poll_out(struct device *dev, unsigned char out_char)
 {
-	/* not supported */
+	struct lpuart_data *data = get_dev_data(dev);
+	bool thread_ctx = !k_is_in_isr() && !k_is_pre_kernel();
+	atomic_val_t o = (atomic_val_t)out_char;
+	int err;
+
+	if (thread_ctx) {
+		/* in thread context pend until tx is in idle */
+		while (data->tx_buf) {
+			k_msleep(1);
+		}
+	} else if (data->tx_buf) {
+		return;
+	}
+
+	if (!atomic_cas(&data->txbyte, -1, o)) {
+		return;
+	}
+
+	err = api_tx(dev, (uint8_t *)&data->txbyte, 1, 1000);
```

Comment:
Maybe timeout value should be set by kconfig?

Reply:
added `NRF_SW_LPUART_DEFAULT_TX_TIMEOUT` which is used in `poll_out` and `fifo_fill`.